src/main/java/stsjorbsmod/actions/ExhumeAndMakeEtherealAction.java

```diff
@@ -0,0 +1,137 @@
+        this.setValues(this.p, AbstractDungeon.player, this.amount);
+        this.actionType = ActionType.CARD_MANIPULATION;
+        this.duration = Settings.ACTION_DUR_FAST;
+    }
+
+    public void update() {
+        Iterator<AbstractCard> c;
+        AbstractCard card;
+        if (this.duration == Settings.ACTION_DUR_FAST) {
+            if (AbstractDungeon.player.hand.size() == 10) {
+                AbstractDungeon.player.createHandIsFullDialog();
+                this.isDone = true;
+
+            } else if (this.p.exhaustPile.isEmpty()) {
+                this.isDone = true;
+
+            } else if (this.p.exhaustPile.size() == 1) {
+                if ((this.p.exhaustPile.group.get(0)).cardID.equals(Exhume.ID) || (this.p.exhaustPile.group.get(0)).cardID.equals(Seance.ID)) {
+                    this.isDone = true;
+                } else {
+
+                    card = this.p.exhaustPile.getTopCard();
+                    card.unfadeOut();
+                    if (!card.isEthereal) {
+                        card.isEthereal = true;
+                        card.rawDescription = prependDescription + card.rawDescription;
+                        card.initializeDescription();
+                    }
+                    this.p.hand.addToHand(card);
+                    if (AbstractDungeon.player.hasPower(CorruptionPower.POWER_ID) && card.type == CardType.SKILL) {
+                        card.setCostForTurn(-9);
+                    }
+                    this.p.exhaustPile.removeCard(card);
+
+                    card.unhover();
+                    card.fadingOut = false;
+                    this.isDone = true;
+                }
+
+            } else {
+                c = this.p.exhaustPile.group.iterator();
+
+                while (c.hasNext()) {
+                    card = c.next();
+                    card.stopGlowing();
+                    card.unhover();
+                    card.unfadeOut();
+                }
+
+                c = this.p.exhaustPile.group.iterator();
+
+
+                while (c.hasNext()) {
+                    card = c.next();
+                    if (card.cardID.equals(Exhume.ID) || card.cardID.equals(Seance.ID)) {
+                        c.remove();
+                        this.exhumes.add(card);
+                    }
+                }
+
+                if (this.p.exhaustPile.isEmpty()) {
+                    this.p.exhaustPile.group.addAll(this.exhumes);
+                    this.exhumes.clear();
+                    this.isDone = true;
+                } else {
+                    AbstractDungeon.gridSelectScreen.open(this.p.exhaustPile, 1, TEXT[0], false);
+                    this.tickDuration();
+                }
+            }
+
+        } else {
+            if (!AbstractDungeon.gridSelectScreen.selectedCards.isEmpty()) {
+                for (c = AbstractDungeon.gridSelectScreen.selectedCards.iterator(); c.hasNext(); card.unhover()) {
```

Comment:
For relatively complex classes like `ExhumeAction`, the decompiler built into IntelliJ isn't very good at producing particularly readable output. I find that [JD-GUI](https://java-decompiler.github.io/) produces output that is much more readable (eg, it doesn't have this confusing `card.unhover()`-as-the-for-loop-incrementor thing in its output, and it also does a better job of decompiling to for-each loops instead of weird iterator nesting patterns); I would recommend starting from that when you need to re-use large parts of a decompiled class

Reply:
I was wondering why the code looked so shoddy! I had no idea the decompiler had such a large impact. I'll be using JD-GUI from now on. 
Do you mind if I leave this one as is for now, considering your idea of patching this into the existing Exhume action?

Reply:
I've rebuilt it anyway from the JD-GUI decompiled ExhumeAction. I've made only the necessary adjustments to it (added Séance to excluded cards, set Ethereal property on card, add string to card description). Oh, and I removed the unused boolean upgrade property from the action.